Update CONTRIBUTING.md to use uv (#41)
kubeflow/sdk: CONTRIBUTING.md

# Contributing to the Kubeflow SDK

This guide explains how to contribute to the Kubeflow SDK project.
For the Kubeflow SDK documentation, please check [the official Kubeflow documentation](https://www.kubeflow.org/docs/components/).

## Requirements
- Python 3.9–3.11
- [pre-commit](https://pre-commit.com/)
- [uv](https://docs.astral.sh/uv/getting-started/installation/)

## Development

The Kubeflow SDK project includes a Makefile with several helpful commands to streamline your development workflow.

Install uv if not installed [Official Docs](https://docs.astral.sh/uv/getting-started/installation/) or using the following command:
```shell
make uv
```
### Install SDK & Dependencies

> Create a virtual environment and install all dependencies (including dev tools):

Reviewer: @szaher Can you fix formatting please: https://github.com/kubeflow/sdk/blob/114fec00669263b6f6302519cd90c570ad8fe4ed/CONTRIBUTING.md#development ?

API Models section also can be removed, since `uv sync --dev` is part of `make install` script.

```sh
make install
```

#### API Models

To install development tools and the latest API modules directly from the master branch:

```shell
uv sync --dev

> ```

Reviewer: Can you delete all of these and just run `make verify` in the Development section ?
Similar to Trainer: https://github.com/kubeflow/trainer/blob/master/CONTRIBUTING.md#development
That should install all required tools and configure uv environment.

Author: actually this one is very specific to installing api modules from master branch but the actual installation of dependencies and app is a couple of lines up.

I added `make install`

```shell
make test-python
```

### E2E Tests
E2E test run in CI on a kind cluster using [Kubeflow Trainer E2E Scripts](https://github.com/kubeflow/trainer/blob/master/CONTRIBUTING.md#e2e-tests).
Clone the `Kubeflow Trainer` repo and run the provided commands against `Trainer` Makefile.
For more details check [the Kubeflow Trainer Contributing Guide](https://github.com/kubeflow/trainer/blob/master/CONTRIBUTING.md#e2e-tests).

## Coding Style
Make sure to install [pre-commit](https://pre-commit.com/) (`uv pip install pre-commit`) and run `pre-commit install` from the root of the repository at least once before creating git commits.

The pre-commit hooks ensure code quality and consistency. They are executed in CI. PRs that fail to comply with the hooks will not be able to pass the corresponding CI gate. The hooks are only executed against staged files unless you run `pre-commit run --all`, in which case, they'll be executed against every file in the repository.

Specific programmatically generated files listed in the `exclude` field in [.pre-commit-config.yaml](.pre-commit-config.yaml) are deliberately excluded from the hooks.

> 

Reviewer 2: Yep, sounds great to me

> 

Author: done

## Requirements
- [Supported Python version](./pyproject.toml#L4)
- [pre-commit](https://pre-commit.com/)
- [uv](https://docs.astral.sh/uv/getting-started/installation/)

## Development

The Kubeflow SDK project includes a Makefile with several helpful commands to streamline your development workflow.

To install all dependencies (including dev tools) and create virtual environment, run

```sh
make install-dev
```

### Coding Style
Make sure to install [pre-commit](https://pre-commit.com/) (`uv pip install pre-commit`) and run `pre-commit install` from the root of the repository at least once before creating git commits.

The pre-commit hooks ensure code quality and consistency. They are executed in CI. PRs that fail to comply with the hooks will not be able to pass the corresponding CI gate. The hooks are only executed against staged files unless you run `pre-commit run --all`, in which case, they'll be executed against every file in the repository.

Specific programmatically generated files listed in the `exclude` field in [.pre-commit-config.yaml](.pre-commit-config.yaml) are deliberately excluded from the hooks.

To check formatting:

```shell
make verify 
```

> #### Pre-commit
> We use pre-commit to ensure consistent code formatting. To enable pre-commit hooks, run:

Reviewer: ```suggestion
```

> ```shell
> pre-commit install
> ```
> To run all hooks manually:
> ```sh
> pre-commit run --all-files

Reviewer: You can remove those sections, since you explain it here: https://github.com/kubeflow/sdk/blob/afe4c9160412090c885e9ab5c214d34c6ea788c3/CONTRIBUTING.md#coding-style

```suggestion
```

> ## Getting Started

Reviewer 2: This is empty, I think we can remove it 
```suggestion
```

## Prerequisites
- Python 3.9–3.11

> - [pip](https://pip.pypa.io/en/stable/)

Reviewer 2: pip is redundant if using uv
```suggestion
```

- [pre-commit](https://pre-commit.com/)
- [uv](https://docs.astral.sh/uv/getting-started/installation/)

## Development
Clone the repository:
```sh
git clone https://github.com/kubeflow/sdk.git
cd sdk
```

Install uv if not installed [Official Docs](https://docs.astral.sh/uv/getting-started/installation/) or using the following command
```sh
make uv
```
### Install SDK & Dependencies
Use uv to create a virtualenv if not created and install dependencies
```sh
uv sync
```

#### Development Build (Optional)
To install development tools and the latest API modules directly from the master branch:
```sh
uv sync

```

## Development Workflow

### Pre-commit

> We use pre-commit to ensure consistent code formatting. To enable pre-commit hooks, run:

Reviewer 2: ```suggestion
We use pre-commit to ensure consistent code formatting and catch issues early. 
Install the hooks:
```

```sh
uv run pre-commit install
```
To run all hooks manually:
```sh

> uv run pre-commit run --all-files

Author: @andreyvelich this is the pre-commit command here!

```

## Testing

> To run the unit tests (if present), execute:

Reviewer 2: ```suggestion
To run the unit tests:
```

```sh

> pytest

Reviewer 2: ```suggestion
uv run pytest
```

## Getting Started

### Prerequisites

> - Python 3.9–3.11

Author: I believe we still need to mention the supported python version if the user has non-supported or old python installation.

> - Python 3.9–3.11

Reviewer: @szaher Did you push this commit ?

> - Python 3.9–3.11

Author: I think it's already there https://github.com/kubeflow/sdk/blob/main/pyproject.toml#L4

> - [pip](https://pip.pypa.io/en/stable/)

Reviewer: `pip` is no longer needed since we have `uv`, isn't ?

- [pre-commit](https://pre-commit.com/)

> - uv

Reviewer: ```suggestion
- [uv](https://docs.astral.sh/uv/getting-started/installation/)
```

Reviewer: @szaher Did you push this commit ?

cd sdk
```

Install uv if not installed [Official Docs](https://docs.astral.sh/uv/getting-started/installation/) or using the following command
```sh

> make uv

Author: No, `make uv` uses the official script for installing uv so it installs it in `~/.local/bin`

cd sdk

> ```

Author: I don't know what is this? It's not part of my change. do you mean 

`cd sdk`
?

coverage report -m
```

## Coding Style

> To check formatting:

Reviewer 2: ```suggestion
To check formatting and linting:
```

```shell
make verify 
```

> 
> ### Using Ruff
> 

Reviewer 2: ```suggestion

To check and auto-fix linting issues:
```

```shell
uv run ruff check --fix
```

> To auto-format, lint all files:
> 

Reviewer 2: ```suggestion
To auto-format all files:

```

### Code Formatting
To check formatting:
```shell
make verify 
```

#### Using Ruff

```shell
uvx ruff check --show-fixes
```

To auto-format, lint all files:

```shell

> uvx ruff check --fix

Author: done

### Code Formatting
To check formatting:
```shell
make verify 
```

#### Using Ruff

```shell
uv run ruff check --show-fixes
```

To auto-format, lint all files:

```shell

> uv run ruff check --fix

Reviewer: In that case, why do we need to run it with `uv run ruff`? After users configure pre-commit hooks, it should automatically format their code.
